packager/packager.cc

```diff
@@ -725,46 +795,122 @@ class ImportTree final {
+                // To save on memory usage by reducing the number of nodes created in the AST, we individually alias
+                // exported constants from the imported package only when there would be an import conflict otherwise -
+                // i.e., when this is either a subpackage of the imported name or imports a subpackage of the imported
+                // name.
+                const bool isOrImportsSubpackage =
+                    absl::c_any_of(package.importedPackageNames,
+                                   [&](const auto &otherImport) -> bool {
+                                       return otherImport.name.fullName.isSuffix(imported.fullName);
+                                   }) ||
+                    package.name.fullName.isSuffix(imported.fullName);
+                mergeImportsFromImportedPackage(ctx, PackageInfoImpl::from(importedPackage), import,
+                                                isOrImportsSubpackage);
             }
         }
     }
 
     // Make add imports for the test package for all normal code exported by its corresponding
     // "normal" package.
-    void mergeSelfExportsForTest(core::Context ctx, const PackageInfoImpl &pkg) {
+    void mergeSelfExports(core::Context ctx, const PackageInfoImpl &pkg, ExportType type) {
         for (const auto &exp : pkg.exports) {
+            if (exp.type != type)
+                continue;
+
             const auto &parts = exp.parts();
             ENFORCE(parts.size() > 0);
-            if (!isTestNamespace(ctx, parts[0])) { // Only add imports for non-test
-                auto loc = exp.fqn.loc.offsets();
-                addImport(pkg, loc, exp.fqn, ImportType::Test);
-            }
+            auto loc = exp.fqn.loc.offsets();
+            addImport(ctx, pkg, loc, exp.fqn, ImportType::Self, true);
         }
     }
 
-    ast::ClassDef::RHS_store makeModule(core::Context ctx, ImportType moduleType) {
+    ast::ClassDef::RHS_store makeModule(core::Context ctx, ModuleType moduleType) {
         vector<core::NameRef> parts;
         ast::ClassDef::RHS_store modRhs;
         makeModule(ctx, &root, parts, modRhs, moduleType, ImportTree::Source());
         return modRhs;
     }
 
 private:
-    void addImport(const PackageInfoImpl &importedPackage, core::LocOffsets loc, const FullyQualifiedName &exportFqn,
-                   ImportType importType) {
+    const bool isNonTestModule(ModuleType moduleType) const {
+        return moduleType == ModuleType::ExportMappingNormal || moduleType == ModuleType::Normal;
+    }
+
+    const bool isExportMappingModule(ModuleType moduleType) const {
+        return moduleType == ModuleType::ExportMappingNormal || moduleType == ModuleType::ExportMappingTest;
+    }
+
+    void mergeImportsFromImportedPackage(core::Context ctx, const PackageInfoImpl &importedPackage,
+                                         const Import &import, bool enumerateExportsInFull) {
+        if (enumerateExportsInFull) {
```

Comment:
This method has a `bool` parameter, which is uses to choose between two totally separate code paths with nothing in common. I think this should probably just be two different methods, with one implementing the `enumerateExportsInFull` branch and the other implementing the only-top-level branch.

Reply:
I think they do have something in common — which is the act of merging exports from an imported package — they differ on what’s being merged.

That said, I can see them being separate methods. Will split them out.